dll/win32/msgina/shutdown.c

```diff
@@ -538,6 +561,116 @@ ShutdownDialog(
     return ret;
 }
 
+/*************************************************************************
+ * SetRegistryToBootInNTNativeMode
+ *
+ * NOTES
+ *     Used to append the NT Native shell start-up arguments to the
+ *     BootExecute registry value so that the system will boot into it.
+ */
+VOID SetRegistryToBootInNTNativeMode(VOID)
+{
+    #if 1
+        /* FIXME: Add proper shell arguments once a native shell is included.
+                  Until then, don't pollute the registry.
+        */
+        return;
+    #endif
+
+    LONG ReturnCode;
+    HKEY HiveKey;
+    DWORD RegValueType = 0;
+    DWORD InitialRegValueSize = 0;
+    DWORD FinalRegValueSize = 0;
+    WCHAR* InitialRegValue = NULL;
+    WCHAR* FinalRegValue = NULL;
+    WCHAR NativeShellArguments[] = L"native Hello World!\0";
+    const UINT ShellArgsLength = lstrlen(NativeShellArguments) + 2;
+
+    ReturnCode = RegCreateKeyExW(
+        HKEY_LOCAL_MACHINE, L"SYSTEM\\CurrentControlSet\\Control\\Session Manager", 0, NULL, REG_OPTION_NON_VOLATILE,
+        KEY_ALL_ACCESS, NULL, &HiveKey, NULL);
+
+    if (ReturnCode != ERROR_SUCCESS)
+    {
+        TRACE("Unable to open the Session Manager key, error %d", GetLastError());
+        return;
+    }
+
+    ReturnCode = RegQueryValueEx(HiveKey, L"BootExecute", NULL, &RegValueType, NULL, &InitialRegValueSize);
+
+    if (ReturnCode != ERROR_SUCCESS || RegValueType != REG_MULTI_SZ)
+    {
+        TRACE("Unable to grab BootExecute, error %d", GetLastError());
+        return;
+    }
+
+    const UINT InitialRegValueLength = InitialRegValueSize / sizeof(WCHAR);
+
+    /* BootExecute is empty (there should be at least autochk in there) - add our string and get out */
+    if (InitialRegValueSize == 0)
+    {
+        FinalRegValueSize = ShellArgsLength * sizeof(WCHAR) + sizeof(WCHAR);
+        FinalRegValue = (WCHAR*)HeapAlloc(GetProcessHeap(), 0, FinalRegValueSize);
+        if (!FinalRegValue)
+        {
+            TRACE("HeapAlloc failed to allocate %d bytes\n", FinalRegValueSize);
+            goto Cleanup;
+        }
+
+        memset(FinalRegValue, '\0', (FinalRegValueSize / sizeof(WCHAR)));
+        memcpy(FinalRegValue, NativeShellArguments, ShellArgsLength * sizeof(WCHAR));
+        goto Cleanup;
+    }
+
+    InitialRegValue = (WCHAR*)HeapAlloc(GetProcessHeap(), 0, InitialRegValueSize);
+    if (!InitialRegValue)
+    {
+        TRACE("HeapAlloc failed to allocate %d bytes", InitialRegValueSize);
+        return;
+    }
+    memset(InitialRegValue, '\0', InitialRegValueLength);
```

Comment:
This needs a byte count, you're passing a character count (but you should really just use `HEAP_ZERO_MEMORY`)